src/pages/best-practices/extensions/security.md

```diff
@@ -0,0 +1,123 @@
+
+<InlineAlert variant="warning" slots="text"/>
+
+Building and executing custom queries with the data adapter does not automatically make them secure. Always use sanitization methods on dynamic data in your queries.
+
+## Use Primary Key
+
+A Primary Key is required for any DB cluster to run effectively. Without a Primary Key, you _will_ see performance issues during table replication.
+
+## Use API for filesystem operations
+
+With the introduction of Remote Storage compatibility, there is no guarantee that files are present in the local filesystem. Because PHP filesystem operations do not support remote storage solutions such as AWS S3, you should always use the Filesystem API to work with the filesystem.
+
+For example, the PHP native function [file_get_contents()](https://www.php.net/manual/en/function.file-get-contents.php) does not allow passing any credentials to authenticate to a remote storage. This functionality might be broken if the source file is located in remote storage.
+
+Use `\Magento\Framework\Filesystem\File\Read::readAll()` method instead. You can also check the list of unsupported PHP methods in the [Magento Coding Standard](https://github.com/magento/magento-coding-standard/blob/develop/Magento2/Sniffs/Functions/DiscouragedFunctionSniff.php) repository.
+
+## Use well-defined indexes
+
+Foreign keys should have indexes. If you are using a field in a WHERE clause of an SQL query you should have an index on it. Such indexes should cover multiple columns based on the queries needed. As a general rule of thumb, indexes should be applied to any column named in the WHERE clause of a SELECT query.
+
+For example, assume we have a user table with a numeric ID (the primary key) and an email address. During log on, MySQL must locate the correct ID by searching for an email. With an index, MySQL can use a fast search algorithm to locate the email almost instantly. Without an index, MySQL must check every record in sequence until the address is found.
+
+It is tempting to add indexes to every column, however, they are regenerated during every table INSERT or UPDATE. That can hit  performance; only add indexes when necessary.
+
+## Avoid using global events
+
+Only on rare occasions would it be necessary to use a global [event](https://glossary.magento.com/event). You should use [frontend](https://glossary.magento.com/frontend) or [adminhtml](https://glossary.magento.com/adminhtml) to narrow the scope instead.
+
+## Use Magento data collections
+
+Execution of a SQL query is one of the most resource-taxing operations. Running SQL queries in a loop often results in a performance bottleneck. To load the EAV model, several heavy queries are required to execute. As the number of executed queries is multiplied with the number of categories, the result is extremely inefficient and slow code. Instead of loading models in a loop, Magento data collections can help to load a set of models in a very efficient manner.
```

Comment:
```suggestion
Execution of a SQL query is one of the most resource-taxing operations. Running SQL queries in a loop often results in a performance bottleneck. To load the EAV model, several heavy queries are required to execute. As the number of executed queries is multiplied with the number of categories, the result is extremely inefficient and slow code. Instead of loading models in a loop, data collections can help to load a set of models in a very efficient manner.
```